src/relax/transform/convert_layout.cc

```diff
@@ -0,0 +1,239 @@
+      }
+    }
+  }
+
+ private:
+  Array<Integer> LayoutToIntegers(const Layout& layout) {
+    Array<Integer> ret;
+    Layout src = InitialLayout(layout.ndim());
+    for (size_t i = 0; i < layout.ndim(); ++i) {
+      ret.push_back(Integer(src.IndexOf(layout[i])));
+    }
+    return ret;
+  }
+
+  Expr RewriteExpr(const Expr& expr, const NLayout& to) {
+    auto fvisitleaf = [&](const Expr& expr, std::array<NLayout, 2> layouts) -> Expr {
+      NLayout from = layouts[0], to = layouts[1];
+      if (NLayoutEqual()(from, to)) return expr;
+      const auto* tensor = GetStructInfoAs<TensorStructInfoNode>(expr);
+      ICHECK(tensor != nullptr) << "Expect a tensor, but got: " << expr;
+      Layout axes = TransposeLike(InitialLayout(tensor->ndim), from.LeafValue(), to.LeafValue());
+      return permute_dims(expr, LayoutToIntegers(axes));
+    };
+    return TransformTupleLeaf<Layout>(
+        VarReplacer::Replace(expr, var_remap_),
+        std::array<NLayout, 2>({GetNLayout(var_layout_map_, expr), to}), fvisitleaf);
+  }
+
+  Array<Expr> RewriteArgs(const Array<Expr>& args, const Array<NLayout>& to) {
+    ICHECK(args.size() == to.size());
+    std::vector<Expr> new_args;
+    for (size_t i = 0; i < args.size(); ++i) {
+      if (IsNestedTensor(args[i])) {
+        new_args.push_back(RewriteExpr(args[i], to[i]));
+      } else {
+        new_args.push_back(args[i]);
+      }
+    }
+    return std::move(new_args);
+  }
+
+  void VisitBinding(const Binding& binding) final {
+    // Emit the binding
+    ExprMutator::VisitBinding(binding);
+    // The layout is default to be initial if not rewritten.
+    if (IsNestedTensor(binding->var)) {
+      if (var_layout_map_.find(binding->var) == var_layout_map_.end()) {
+        var_layout_map_[binding->var] = InitialNLayout(binding->var);
+      }
+    }
+  }
+
+  Expr VisitVars_(const Var& var) {
+    if (IsNestedTensor(var)) {
+      return RewriteExpr(var, InitialNLayout(var));
+    }
+    return ExprMutator::VisitExpr_(var.get());
+  }
+
+  Expr VisitExpr_(const VarNode* op) final { return VisitVars_(GetRef<Var>(op)); }
+
+  Expr VisitExpr_(const DataflowVarNode* op) final { return VisitVars_(GetRef<Var>(op)); }
+
+  static Optional<InferLayoutOutput> GetInferLayoutInfo(const CallNode* call_node,
+                                                        Map<String, Array<String>> desired_layouts,
+                                                        VarLayoutMap var_layout_map) {
+    const OpNode* op_node = call_node->op.as<OpNode>();
+    if (op_node == nullptr) return NullOpt;
+    Op op = Downcast<Op>(GetRef<Op>(op_node));
+    const auto attr_map = Op::GetAttrMap<FRelaxInferLayout>("FRelaxInferLayout");
+    if (attr_map.count(op)) {
+      FRelaxInferLayout f = attr_map[op];
+      return f(GetRef<Call>(call_node), desired_layouts, var_layout_map);
+    } else {
+      return NullOpt;
+    }
+  }
+
+  void VisitBinding_(const VarBindingNode* binding, const CallNode* call_node) override {
```

Comment:
Let’s use `final` for overridden methods.
```suggestion
  void VisitBinding_(const VarBindingNode* binding, const CallNode* call_node) final {
```